src/connection/connection-manager.ts

```diff
@@ -23,13 +23,12 @@ export class InMemoryConnectionManager implements ConnectionManager {
   constructor(private dwn: Dwn, private connections: Map<WebSocket, SocketConnection> = new Map()) {}
 
   async connect(socket: WebSocket): Promise<void> {
-    const connection = new SocketConnection(socket, this.dwn);
-    this.connections.set(socket, connection);
-    // attach to the socket's close handler to clean up this connection.
-    socket.on('close', () => {
-      // the connection internally already cleans itself up upon a socket close event, we just ned to remove it from our set.
+    const connection = new SocketConnection(socket, this.dwn, () => {
```

Comment:
For my learning: was this refactoring triggered by some gotcha that I should be aware of?

Reply:
A bit of clean up/segregation of concerns that was found by a gotcha.

In `SocketConnection.close()` I call `this.socket.removeAllListeners()` towards the top of the function. This also removes the `socket.on` listener that was called outside of `SocketConnection`.

I could have just removed listeners for 'message' first, which was my intent  in order to stop new messages from triggering events immediately. And then I could have cleaned up the other listeners at the end.

But when I ran into this I really thought that `SocketConnection` should interface with the socket handlers and expose a clean up method and that would be a better separation of concerns anyway.

Reply:
Didn't fully get it but sounds like a move in the better direction!